SQL-1502: Native Query (#31)
mongodb/mongo-powerbi-connector: testing.md

# Manual Testing (for Native Query)

At this time, Native Query must be tested manually by reviewers.

The following query is sufficient to test Native Query as it exercises mutliple areas of translation.

1. Ensure you have a local ADF running.

> - Navigate to the `mongo-odbc-driver` repository locally.

Reviewer: In the past we've tried to keep testing resources in the same repo instead of relying on one of our others.  Since we have a run_adf.sh script in this mongo-powerbi-connector repo, how about we use one of the collections that we're already using here like [sales](https://github.com/mongodb/mongo-powerbi-connector/blob/main/resources/integration_test/testdata/sales.json).  Or alternatively add a few rows of the grades and class collections from the mongo-odbc-driver integration tests. 
For loading you can use mongoimport, add the steps:
```
          MONGO_TOOLS_VERSION=100.6.1
          curl -L -o mongodb-tools.zip \
            https://fastdl.mongodb.org/tools/db/mongodb-database-tools-windows-x86_64-$MONGO_TOOLS_VERSION.zip
          unzip -jn mongodb-tools.zip
          chmod +x mongoimport.exe
 ```

```
          ./mongoimport.exe --uri="mongodb://localhost:28017/supplies" \
            --drop resources/integration_test/testdata/sales.json
```

```
MONGOSH=$(find ./local_adf/ | grep mongo.exe | head -1)

# Replace with the namespace you are adding
$MONGOSH  -u mhuser -p pencil --eval 'db.runCommand({sqlGenerateSchema: 1, sampleNamespaces: ["supplies.sales"], setSchemas: true})' localhost/admin
```

Author: Sounds good, I can make the change. In general, what do you think about converging the data we use to test across our different projects?

Reviewer: I am okay with not converging the data used for the different projects.  I think the datasets are usually chosen for testing specific capabilities of the drivers and might not be super useful to have the same dataset for all repos.

- Run the ADF script via `./resources/run_adf start`
- After it sets up and returns back to the shell prompt, load local data via `cargo run --bin data_loader`
- Navigate back to the `mongo-powerbi-connector` repository.

2. Build the connector. The easiest way to do this is in VSCode, using the Power Query SDk.
- Navigate to the `connector` directory.
- `code .` to open the project in VSCode.
- Open `MongoDBAtlasODC.pq`

> - Right click, and choose *Evaluate current power query file*

Reviewer: ```suggestion
- On the editor window, right click on  `MongoDBAtlasODC.pq`, and choose *Evaluate current power query file*
```
At least with my VSCode I have to click on the actual editor window, this might help if someone tries clicking on the filename

Author: good call

> - Copy `bin\AnyCPU\Debug\connector.mez` to the Microsoft Power BI custom connector folder, located in `C:\Users\<username>\Documents\Microsoft Power BI Desktop\Custom Connectors`.

Reviewer: ```suggestion
- Copy `bin\AnyCPU\Debug\connector.mez` to the Microsoft Power BI custom connector folder, located in `C:\Users\<username>\Documents\Power BI Desktop\Custom Connectors`.
```

Is your Power BI version the custom one?  My custom connector goes into: `C:\Users\[username]\Documents\Power BI Desktop\Custom Connectors`

> Put the connector .pq, .pqx, .m, or .mez file in your local [Documents]\Power BI Desktop\Custom Connectors folder. If the folder doesn't exist, create it.
https://learn.microsoft.com/en-us/power-bi/connect-data/desktop-connector-extensibility#custom-connectors

Author: This is how I've always had it. I wonder if they allow both, as I seem to remember their docs specifically called out the name as `Microsoft Power BI Desktop`. I have both the versions you download from the internet or that you can install via the windows store.

I searched some on this and it appears that it could be named both, but don't have both. https://community.fabric.microsoft.com/t5/Developer/Custom-Connectors-Path/m-p/706338#M19564

I can update this if you want.

Reviewer: Oh, Yeah let's update it since their official doc says `Power BI Desktop`.

# Manual Testing (for Native Query)

At this time, Native Query must be tested manually by reviewers.

The following query is sufficient to test Native Query as it exercises mutliple areas of translation.

1. Ensure you have a local ADF running.

> - Navigate to the `mongo-odbc-driver` repository locally.

Reviewer: ```suggestion
- Navigate to the `mongo-powerbi-connector` repository locally.
```
Updating the repo to powerbi

- Run the ADF script via `./resources/run_adf start`
- Load data
  - If you have `mongoimport.exe` already installed, you can skip the downloading it. Otherwise
    ```
    MONGO_TOOLS_VERSION=100.6.1 curl -L -o mongodb-tools.zip \
    https://fastdl.mongodb.org/tools/db/mongodb-database-tools-windows-x86_64-$MONGO_TOOLS_VERSION.zip
    unzip -jn mongodb-tools.zip
    chmod +x mongoimport.exe
    ```
  - Import the data
    ```
    ./mongoimport.exe --uri="mongodb://localhost:28017/supplies" --drop resources/integration_test/testdata/sales.json
    ```
  - Set the ADF schema
    ```
    MONGOSH=$(find ./local_adf/ | grep mongo.exe | head -1)
    # Replace with the namespace you are adding
    $MONGOSH  -u mhuser -p pencil --eval 'db.runCommand({sqlGenerateSchema: 1, sampleNamespaces: ["supplies.sales"], setSchemas: true})' localhost/admin
    ```

> - Navigate back to the `mongo-powerbi-connector` repository.

Reviewer: ```suggestion
```

2. Build the connector. The easiest way to do this is in VSCode, using the Power Query SDk.
- Navigate to the `connector` directory.
- `code .` to open the project in VSCode.
- Open `MongoDBAtlasODBC.pq`
- On the editor window, right click anywhere in the editor window for the `MongoDBAtlasODBC.pq` file and choose *Evaluate current power query file*
- Copy `bin\AnyCPU\Debug\connector.mez` to the Microsoft Power BI custom connector folder, located in `C:\Users\<username>\Documents\Power BI Desktop\Custom Connectors`.

3. Start Power BI.
4. Choose *Get Data*
5. Search for "mongodb" in the search box, and select the *MongoDB Atlas SQL (Beta)* connector.
6. Connect to your local ADF. 
- The URI should be `mongodb://localhost` 
- The database is `integration_test`.
- In the Native Query box, input the following SQL query which groups all sold items by name and calculates out how much revenue they generate per sale.
  ```

>   SELECT name, AVG(price * quantity) as revenue_per_sale FROM(SELECT items_name as name, items_price as price, items_quantity as quantity FROM FLATTEN(UNWIND(sales WITH PATH => items))) as derived GROUP BY name

Reviewer: Great work on finding a suitable SQL query and the equivalent MQL 🚀
